runtime/observability/telemetry-provider-micrometer/jvm/src/aws/smithy/kotlin/runtime/telemetry/micrometer/MicrometerMeterProvider.kt

```diff
@@ -0,0 +1,211 @@
+    private val meter = MicrometerMeter(meterRegistry)
+
+    override fun getOrCreateMeter(scope: String): Meter = meter
+}
+
+private class MicrometerMeter(
+    private val meterRegistry: MeterRegistry,
+) : Meter {
+    override fun createUpDownCounter(name: String, units: String?, description: String?): UpDownCounter =
+        MicrometerUpDownCounter(
+            meterMetadata = MeterMetadata(name, units, description),
+            meterRegistry = meterRegistry,
+        )
+
+    override fun createAsyncUpDownCounter(
+        name: String,
+        callback: LongUpDownCounterCallback,
+        units: String?,
+        description: String?,
+    ): AsyncMeasurementHandle =
+        MicrometerLongGauge(
+            callback = callback,
+            meterMetadata = MeterMetadata(name, units, description),
+            meterRegistry = meterRegistry,
+        )
+
+    override fun createMonotonicCounter(name: String, units: String?, description: String?): MonotonicCounter =
+        MicrometerMonotonicCounter(
+            meterMetadata = MeterMetadata(name, units, description),
+            meterRegistry = meterRegistry,
+        )
+
+    override fun createLongHistogram(name: String, units: String?, description: String?): LongHistogram =
+        MicrometerLongHistogram(
+            MicrometerDoubleHistogram(
+                meterMetadata = MeterMetadata(name, units, description),
+                meterRegistry = meterRegistry,
+            )
+        )
+
+    override fun createDoubleHistogram(name: String, units: String?, description: String?): DoubleHistogram =
+        MicrometerDoubleHistogram(
+            meterMetadata = MeterMetadata(name, units, description),
+            meterRegistry = meterRegistry,
+        )
+
+    override fun createLongGauge(
+        name: String,
+        callback: LongGaugeCallback,
+        units: String?,
+        description: String?,
+    ): AsyncMeasurementHandle = MicrometerLongGauge(
+        callback = callback,
+        meterMetadata = MeterMetadata(name, units, description),
+        meterRegistry = meterRegistry,
+    )
+
+    override fun createDoubleGauge(
+        name: String,
+        callback: DoubleGaugeCallback,
+        units: String?,
+        description: String?,
+    ): AsyncMeasurementHandle = MicrometerDoubleGauge(
+        callback = callback,
+        meterMetadata = MeterMetadata(name, units, description),
+        meterRegistry = meterRegistry,
+    )
+
+}
+
+private data class MeterMetadata(val meterName: String, val units: String?, val description: String?)
+
+private class MicrometerUpDownCounter(
+    private val meterMetadata: MeterMetadata,
+    private val meterRegistry: MeterRegistry,
+) : UpDownCounter {
+    override fun add(value: Long, attributes: Attributes, context: Context?) {
+        meterMetadata
+            .counter()
+            .tags(attributes.toTags())
+            .register(meterRegistry)
```

Comment:
> Does registering a counter with an identical name multiple times cause a problem?

It doesn't. If a counter with the given name and set of tags already exists in MicroMeter, it will simply return the existing instance.